src/test/java/com/thomashan/collection/immutable/ImmutableListImplTest.java

```diff
@@ -0,0 +1,254 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ImmutableListImplTest {
+    @Test
+    public void testSize() {
+        assertEquals(0, of().size());
+    }
+
+    @Test
+    public void testIsEmpty_GivenEmptyList_ReturnsTrue() {
+        assertTrue(of().isEmpty());
+    }
+
+    @Test
+    public void testIsEmpty_GivenNonEmptyList_ReturnsFalse() {
+        assertFalse(of(new Object()).isEmpty());
+    }
+
+    @Test
+    public void testContains_GivenObjectIsInList_ReturnsTrue() {
+        Object object = new Object();
+
+        assertTrue(of(object).contains(object));
+    }
+
+    @Test
+    public void testContains_GivenObjectIsNotInList_ReturnsFalse() {
+        Object object = new Object();
+
+        assertFalse(of(object).contains(new Object()));
+    }
+
+    @Test
+    public void testIterator_ReturnsIterator() {
+        assertTrue(of().iterator() instanceof Iterator);
+    }
+
+    @Test
+    public void testToArray() {
+        assertArrayEquals(new Object[0], of().toArray());
+    }
+
+    @Test
+    public void testToArray_GivenIntegerArray_ReturnIntegerArray() {
+        assertArrayEquals(new Integer[0], of().toArray(new Integer[0]));
+    }
+
+    @Test
+    public void testAdd_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of().add(new Object()));
+    }
+
+    @Test
+    public void testRemove_ThrowsUnsupportedOperationException() {
+        Object object = new Object();
+
+        assertThrows(UnsupportedOperationException.class, () -> of(object).remove(object));
+    }
+
+    @Test
+    public void testContainsAll_GivenAllObjectAreInTheList_ReturnsTrue() {
+        Object object = new Object();
+
+        assertTrue(of(object).containsAll(Collections.singleton(object)));
+    }
+
+    @Test
+    public void testContainsAll_GivenAllObjectNotInTheList_ReturnsFalse() {
+        assertFalse(of(new Object()).containsAll(Collections.singleton(new Object())));
+    }
+
+    @Test
+    public void testAddAll_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of().addAll(Collections.singleton(new Object())));
+    }
+
+    @Test
+    public void testAddAllWithIndex_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of().addAll(0, Collections.singleton(new Object())));
+    }
+
+    @Test
+    public void testRemoveAll_ThrowsUnsupportedOperationException() {
+        Object object = new Object();
+
+        assertThrows(UnsupportedOperationException.class, () -> of(object).removeAll(Collections.singleton(object)));
+    }
+
+    @Test
+    public void testRetainAll_ThrowsUnsupportedOperationException() {
+        Object object = new Object();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Found 'DU'-anomaly for variable 'object' (lines '104'-'107').](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4509131)